### PR TITLE
Add attention and final logit soft-capping, update scaling factor to Gemma2

### PR DESCRIPTION
**Pull request by abetlen**
This PR adds the missing attention layer and final logit soft-capping. Implementation referenced from huggingface transformers. Additionally Gemma2 applies a pre-attention scaling of `hidden_size / num_attention_heads`.

**NOTE**: attention soft-capping is not compatible with flash attention so flash attention is disabled when loading the model. 

Once this PR is finalised / merged the gguf will need to be generated again to include the soft-capping scales.

@slaren let me kv names / hparams should be changed or if anything stands out to you.

- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

### CODE REVIEW

File: convert-hf-to-gguf.py [hunk 1]
```diff
@@ -2363,6 +2363,14 @@ def set_gguf_parameters(self):
...
         self.gguf_writer.add_value_length(hparams["head_dim"])
         self.gguf_writer.add_file_type(self.ftype)
+        self.gguf_writer.add_float32(
+            gguf.Keys.LLM.ATTN_LOGIT_SOFTCAPPING.format(arch=self.model_arch),
+            self.hparams["attn_logit_softcapping"]
+        )
+        self.gguf_writer.add_float32(
+            gguf.Keys.LLM.FINAL_LOGIT_SOFTCAPPING.format(arch=self.model_arch),
+            self.hparams["final_logit_softcapping"]
+        )
```
Comment (slaren) [inline convert-hf-to-gguf.py:2373]: I think these need custom `add_` functions like the other parameters.

File: src/llama.cpp [hunk 2]
```diff
@@ -7579,6 +7590,12 @@ static struct ggml_tensor * llm_build_kqv(
             kq = ggml_scale(ctx, kq, 30);
         }
 
+        if (hparams.attn_soft_cap) {
+            kq = ggml_scale(ctx, kq, 1.0f / hparams.f_attn_logit_softcapping);
+            kq = ggml_tanh(ctx, kq);
+            kq = ggml_scale(ctx, kq, hparams.f_attn_logit_softcapping);
+        }
```
Comment (slaren) [inline src/llama.cpp:7597]: Since this is not compatible with flash attention, it should be disabled in `llama_new_context_with_model`. There are other similar cases where it is disabled there.

**Comment by N8python** [comment]
This is absolutely awesome!

File: convert-hf-to-gguf.py (hunk 1 above)
Comment (slaren) [reply convert-hf-to-gguf.py]: I tried converting the 9b base model and the keys got these names:
```
lama_model_loader: - kv  12:                  27.attn_logit_softcapping f32              = 50.000000
llama_model_loader: - kv  13:                 27.final_logit_softcapping f32              = 30.000000
```
So this isn't being formatted correctly, it is using the arch index instead of the name.

File: src/llama.cpp [hunk 3]
```diff
@@ -17379,6 +17402,11 @@ struct llama_context * llama_new_context_with_model(
         params.flash_attn = false;
     }
 
+    if (params.flash_attn && model->arch == LLM_ARCH_GEMMA2) {
+        LLAMA_LOG_WARN("%s: flash_attn is not compatible with Gemma2 - forcing off\n", __func__);
+        params.flash_attn = false;
+    }
```
Comment (slaren) [inline src/llama.cpp:17408]: This should check for `attn_soft_cap` instead of the arch. Eg:
```suggestion
    if (params.flash_attn && model->hparams.attn_soft_cap) {
        LLAMA_LOG_WARN("%s: flash_attn is not compatible with attn_soft_cap - forcing off\n", __func__);
        params.flash_attn = false;
    }

```

**Comment by N8python** [comment]
Will this require updating the commit of llama.cpp, or just the gguf?

File: convert-hf-to-gguf.py (hunk 1 above)
Comment (slaren) [reply convert-hf-to-gguf.py]: It's working now.

**Comment by abetlen** [comment]
> Will this require updating the commit of llama.cpp, or just the gguf?

Both but I think this is the final set of changes.

**Comment by slaren** [comment]
Perplexity is looking much better now with the 9b base model.
```
[1]9.2071,[2]9.7469,[3]8.5964,[4]8.9120,[5]9.0974,[6]9.5314,[7]9.5795,[8]9.9935,[9]10.8124,[10]11.5707,[11]11.2326,[12]11.6694,[13]12.1546,[14]11.2626,[15]10.7205,[16]10.8900,[17]10.2102,[18]10.4680,[19]10.3165,[20]10.1336,[21]9.9171,[22]9.8545,[23]9.3609,[24]8.9963,[25]8.7959,[26]8.4272,[27]8.2837,[28]8.2287,[29]8.1524,[30]8.1909,[31]8.1299,[32]8.1604,[33]8.2355,[34]8.3548,[35]8.3592,[36]8.4679,[37]8.5151,[38]8.4315,[39]8.4928,[40]8.5288,[41]8.4178,[42]8.4029,[43]8.4344,[44]8.3349,[45]8.2776,[46]8.2448,[47]8.3508,[48]8.3703,[49]8.3698,[50]8.4112,[51]8.3984,[52]8.4121,[53]8.4798,[54]8.4386,[55]8.5013,[56]8.4724,[57]8.4617,[58]8.5479,[59]8.5988,[60]8.6415,[61]8.6199,[62]8.6748,[63]8.7169,[64]8.7935,[65]8.8943,[66]8.9945,[67]8.9497,[68]8.9609,[69]8.9441,[70]8.9776,[71]9.0380,[72]9.0679,[73]9.0876,[74]9.0596,[75]9.0648,[76]9.0820,[77]9.1173,[78]9.0350,[79]9.0445,[80]8.9748,[81]9.0271,[82]9.0024,[83]9.0165,[84]9.0856,[85]9.1867,[86]9.2471,[87]9.2703,[88]9.2412,[89]9.1967,[90]9.2079,[91]9.1894,[92]9.2655,[93]9.2878,[94]9.3000,[95]9.3255,[96]9.3562,[97]9.3499,[98]9.3618,[99]9.4381,[100]9.4716,
Final estimate: PPL = 9.4716 +/- 0.18491
```

**Comment by slaren** [comment]
The chat template continues adding one line on each reply, there is probably something wrong there.

**Review by slaren — APPROVED** [review]

**Comment by theo77186** [comment]
Looks really nice. This would require regenerating gguf files as using existing gguf files don't have the required keys. Attempting to load an old gguf file with this PR results in this error:
```
llama_model_load: error loading model: error loading model hyperparameters: key not found in model: gemma2.attn_logit_softcapping
```

**Comment by abetlen** [comment]
@slaren thank you for checking!

> The chat template continues adding one line on each reply, there is probably something wrong there.


With regards to the newlines I wonder if this is [relevant](https://huggingface.co/google/gemma-7b-it/discussions/38#65d7b14adb51f7c160769fa1)

**Comment by slaren** [comment]
I am using the built-in chat template support, with `-cnv`. The result is this:
```
$ ./llama-cli -m models/gemma-2-9b-it/ggml-model-f16.gguf -ngl 99 -cnv 
[...]
== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to the AI.
 - To return control without starting a new line, end your input with '/'.
 - If you want to submit another line, end your input with '\'.


> say 1
1

> say 2
2

> say 3
3


> say 4
4

> say 5
5


> say 6
6

>
```

Sometimes there is between one or two extra lines, which seems odd.

**Comment by abetlen** [comment]
> Looks really nice. This would require regenerating gguf files as using existing gguf files don't have the required keys. Attempting to load an old gguf file with this PR results in this error:
> 
> ```
> llama_model_load: error loading model: error loading model hyperparameters: key not found in model: gemma2.attn_logit_softcapping
> ```

That's when you're loading an older gemma2 gguf correct? Yes, those will need to be regenerated.

File: src/llama.cpp [hunk 4]
```diff
@@ -2099,6 +2103,9 @@ struct llama_hparams {
     float f_norm_eps;
     float f_norm_rms_eps;
 
+    float f_attn_logit_softcapping;
+    float f_final_logit_softcapping;
```
Comment (ngxson) [inline src/llama.cpp:2107]: Because there are already quite a few gguf circulating in the internet, can we provide a default value so that it doesn't break existing model?

Value taken from https://huggingface.co/google/gemma-2-9b/blob/main/config.json and https://huggingface.co/google/gemma-2-27b/blob/main/config.json

```cpp
    float f_attn_logit_softcapping = 50.0;
    float f_final_logit_softcapping = 30.0;
```

Then we mark the code to get KV as non-required

```cpp
        case LLM_ARCH_GEMMA2:
            {
                ml.get_key(LLM_KV_ATTENTION_LAYERNORM_RMS_EPS, hparams.f_norm_rms_eps);
                ml.get_key(LLM_KV_ATTN_LOGIT_SOFTCAPPING, hparams.f_attn_logit_softcapping, false);
                ml.get_key(LLM_KV_FINAL_LOGIT_SOFTCAPPING, hparams.f_final_logit_softcapping, false);
                hparams.attn_soft_cap = true;
```

**Comment by ngxson** [comment]
@slaren Thanks for spotting that error. I made a typo in the code, will be fixed with #8198

**Comment by ngxson** [comment]
I made a dirty patch to see it that's the root cause or not, but seems like the model still want to output multiple new lines:

```
'':108, '<start_of_turn>':106, 'user':1645, '':108, 'say':23523, ' sunflower':74207, ' in':575, ' emoji':52810, '<end_of_turn>':107, '':108, '<start_of_turn>':106, 'model':2516, '':108, '':241549, '  ':139, '':111, '<end_of_turn>':107
```

Explanation: token 108 `\n` is added before `<start_of_turn>` for user (in other words, after `<end_of_turn>` of the last turn)

But for some reasons, it still output token 111 `\n\n\n\n`, so I assume there is something to do with the model itself. Please note that I'm using `-t 0` for all of me tests.

**Review by ngxson — APPROVED** [review]

**Review by ggerganov — APPROVED** [review]
Regarding the new lines - maybe it's best to compare with the reference inference results and see if these are expected

**Comment by qnixsynapse** [comment]
BTW, Now am thinking that this is also probably needed.

From the technical paper of Gemma2 :
>We alternate between a local sliding window attention (Beltagy et al., 2020a,b) and global attention (Luong et al., 2015) in every other layer. The sliding window size of local attention layers is set to 4096 tokens, while the span of the global attention layers is set to 8192 tokens.

I am not sure if this is implemented or not( I am still weak in C++, but still I am trying to find out in the code).

**Comment by abetlen** [comment]
It seems that there's another difference as well, the attention scaling factor is a custom value of [`1 / sqrt(144)`](https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/gemma2/modeling_gemma2.py#L202).

**Comment by sinand99** [comment]
I hope this fixes Gemma and Phi models which are both broken now. Getting forever-repeating bug after they exceed their context length. Changing n_predict does not work.

**Comment by arlo-phoenix** [comment]
> Edit2: We don't even have a sliding window attention logic in either [llm_build_kv](https://github.com/ggerganov/llama.cpp/blob/72272b83a3878e91251218c981b4c6ec16c33912/src/llama.cpp#L7623) or [llm_build_kv_store](https://github.com/ggerganov/llama.cpp/blob/72272b83a3878e91251218c981b4c6ec16c33912/src/llama.cpp#L7210) functions, nor in the [attention implementation for gemma2](https://github.com/ggerganov/llama.cpp/blob/72272b83a3878e91251218c981b4c6ec16c33912/src/llama.cpp#L11024)... Am I missing something?

No, it's not implemented, but since ggerganov refactored the KQ masking a while ago the change should be simple, the issue for SWA was also just reopened. I already tried implementing it for gemma2, see my comment [here](https://github.com/ggerganov/llama.cpp/issues/3377#issuecomment-2198329330)

**Comment by arlo-phoenix** [comment]
@qnixsynapse hacked it in here: https://github.com/arlo-phoenix/llama.cpp/tree/gemma2 Doesn't need new ggufs to work. Output is better at the start, but then runs into a repeating issue, so something is probably still missing or I messed something up. I'll stop for today though.

**Comment by abetlen** [comment]
I've updated the scaling factor approach to compute the value instead of adding yet another key to the gguf. This value is computed using `hidden_size / num_attention_heads` as in the original pytorch implementation.

**Comment by qnixsynapse** [comment]
> hacked it in here: https://github.com/arlo-phoenix/llama.cpp/tree/gemma2 Doesn't need new ggufs to work. Output is better at the start, but then runs into a repeating issue, so something is probably still missing or I messed something up. I'll stop for today though.

@arlo-phoenix Good work!

I will take a look at it today(Sunday).

Also, just to mention, I am still finding the HF's Transformer's Gemma 2 implementation to be somewhat subpar since the model hosted at HF chat still gives much low quality responses than the model hosted at Google AI studio. Not sure if the transformers library being used there is updated with the latest fixes or not.
So we cannot blindly depend on HF transformers library's implementation for now imo.

**Comment by ddh0** [comment]
Now that this is merged, if I make a new GGUF, am I good to go? Or are there other fixes I should wait for?

**Comment by qnixsynapse** [comment]
@ddh0 Currently, @arlo-phoenix is using hardcoded window size for testing. I think best is to regenerate ggufs once this(alternate SWA) is fixed.

File: src/llama.cpp [hunk 5]
```diff
@@ -11039,7 +11056,7 @@ struct llm_build_context {
                         ext_factor, attn_factor, beta_fast, beta_slow);
                 cb(Qcur, "Qcur", il);
 
-                Qcur = ggml_scale(ctx0, Qcur, 1.0f / sqrtf(float(n_embd_head_k)));
+                Qcur = ggml_scale(ctx0, Qcur, 1.0f / sqrtf(float(n_embd / n_head)));
```
Comment (slaren) [inline src/llama.cpp:11059]: I am not sure that this change is correct, and it will affect other models.

Comment (bartowski1182) [reply src/llama.cpp:11059]: This is inside the build gemma2 graph right? So it may be incorrect but shouldn't affect other models

Comment (slaren) [reply src/llama.cpp:11059]: Yes you are right, I though this part of the common `build_kv` function.

Comment (slaren) [reply src/llama.cpp:11059]: I am not sure if this is correct, this change increases ppl of f16 9b-it @ 100 chunks from 11.9212 to 11.9850.

Comment (ngxson) [reply src/llama.cpp:11059]: Not sure if it's related, but original `num_key_value_heads` is set to 16, while gguf version has `gemma2.attention.head_count_kv = 8`

https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/gemma2/configuration_gemma2.py#L105

Comment (ggerganov) [reply src/llama.cpp:11059]: It seems to be correct, because in the transformers implementation they do the following scaling:

https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/gemma2/modeling_gemma2.py#L256-L258

`self.scaling` is initialized on this line:

https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma2/modeling_gemma2.py#L202

So it uses the `query_pre_attn_scalar` value from the `config.json`:

https://huggingface.co/google/gemma-2-9b-it/blob/93be03fbe3787f19bf03a4b1d3d75d36cb1f6ace/config.json#L24

For the 9B model, this scalar value is `224` which is equal to `n_embd / n_head = 3584 / 16 = 224`, while `n_embd_head_k == 256`. Technically, it will be more correct to read this scaling value and use it directly, but it seems that for all Gemma 2 models so far, this factor is equal to `n_embd / n_head`. Maybe we just need to assert it during Python conversion

File: src/llama.cpp [hunk 6]
```diff
@@ -11106,6 +11123,12 @@ struct llm_build_context {
 
         // lm_head
         cur = ggml_mul_mat(ctx0, model.output, cur);
+
+        // final logit soft-capping
+        cur = ggml_scale(ctx0, cur, 1.0f / hparams.f_final_logit_softcapping);
```
Comment (eran-medan) [inline src/llama.cpp:11128]: @ggerganov 
Total nitpick that probably should be ignored. I came here from curiosity, and I know this is merged by now and I have absolutely no place to comment. But 
Isn’t this similar logic to lines 7594-7596? 

I know the “rule of 3” but shouldn’t this be extracted to apply_softcap method? Or is it too trivial to need that? (Risk is someone down the road might tweak one place and not be aware they need to do so in the other… )
